universal-idl-program/program/src/account_checks.rs

```diff
@@ -0,0 +1,170 @@
+}
+
+#[inline(always)]
+pub fn assert_empty(account: &AccountInfo, name: &str) -> ProgramResult {
+    assert_with_msg(
+        account.data_is_empty(),
+        ProgramError::InvalidInstructionData,
+        lazy_format!("{} must be empty", name),
+    )
+}
+
+#[inline(always)]
+pub fn assert_address(address_one: &Pubkey, address_two: &Pubkey, name: &str) -> ProgramResult {
+    assert_with_msg(
+        address_one == address_two,
+        ProgramError::InvalidInstructionData,
+        lazy_format!("{} must equal {}", name, address_two),
+    )
+}
+
+#[inline(always)]
+pub fn assert_program_account(account: &AccountInfo, discriminator: [u8; 8]) -> ProgramResult {
+    let data = &account.data.borrow_mut();
+    assert_with_msg(
+        is_correct_account_type(data, discriminator) && *account.owner == crate::id(),
+        ProgramError::InvalidInstructionData,
+        lazy_format!("Invalid account type for {}", account.key),
+    )
+}
+
+pub fn assert_account_owned_by_loader(account: &AccountInfo, name: &str) -> ProgramResult {
+    assert_owner(account, &bpf_loader_upgradeable::id(), name)
+}
+
+pub fn assert_system_program(account: &AccountInfo) -> ProgramResult {
+    assert_address(account.key, &system_program::id(), "system_program")
+}
+
+pub fn assert_program_data_matches_program(
+    program: &UpgradeableLoaderState,
+    program_data_key: &Pubkey,
+) -> ProgramResult {
+    if let UpgradeableLoaderState::Program {
+        programdata_address,
+    } = program
+    {
+        assert_with_msg(
+            program_data_key == programdata_address,
+            ProgramError::InvalidInstructionData,
+            lazy_format!("Invalid program data for {}", program_data_key),
+        )?;
+        Ok(())
+    } else {
+        Err(ProgramError::InvalidInstructionData)
+    }
+}
+
+pub fn assert_program_authority_matches_program(
+    program_data: &UpgradeableLoaderState,
+    program_authority_key: &Pubkey,
+) -> ProgramResult {
+    if let UpgradeableLoaderState::ProgramData {
+        upgrade_authority_address,
+        ..
+    } = program_data
+    {
+        assert_with_msg(
+            upgrade_authority_address.is_some(),
+            ProgramError::InvalidInstructionData,
+            lazy_format!("The given program is frozen, and has no authority"),
+        )?;
+
+        assert_with_msg(
+            upgrade_authority_address.unwrap() == *program_authority_key,
+            ProgramError::InvalidInstructionData,
+            lazy_format!("Invalid program authority: {}", program_authority_key),
+        )?;
+        Ok(())
+    } else {
+        Err(ProgramError::InvalidInstructionData)
+    }
+}
+
+pub fn is_correct_account_type(data: &[u8], discriminator: [u8; 8]) -> bool {
+    data[..8] == discriminator
+}
+
+pub fn assert_program_authority_in_allowlist(account: &AccountInfo, name: &str) -> ProgramResult {
+    let allowlist_frozen_authorities: Vec<&str> =
+        ["6tgR1upn2bsMdiprpfUAWmoniEJ8E1XVF8f9gLmWqyTS"].to_vec();
```

Comment:
Interesting approach:

List of admins that are able to be meta_authorities to support IDLs for frozen programs